app/models/project_step.rb

```diff
@@ -135,6 +138,27 @@ def handle_original_date_logic
     end
   end
 
+  def handle_finalized_at
+    if is_finalized && ! finalized_at
+      self.finalized_at = Time.now
+    elsif ! is_finalized && finalized_at
+      self.finalized_at = nil
+    end
+  end
+
+  # Validates that a step may not be unfinalized more than 24 hours since it was previously marked
+  # as finalized.  Note, should generally be avoided by front-end logic, but guards against edge
+  # cases.
+  def unfinalize_allowed
+    if is_finalized_changed? && ! is_finalized && is_finalized_locked?
+      errors.add(:is_finalized, I18n.t("project_step.unfinalize_disallowed"))
+    end
+  end
```

Comment:
A quick unit test for this validation would probably be a good idea. I don't want to get out of the habit of writing unit tests for this project. No need to test every supporting method, but this one would be good. Just testing the positive and negative cases.